File: Python/polygon.py
from typing import Iterable


class Point:
    def __init__(self, x: float, y: float) -> None:
        self.coords = (x, y)

    @property
    def x(self):
        return self.coords[0]

    @property
    def y(self):
        return self.coords[1]


class Edge:
    def __init__(self, start: Point, end: Point) -> None:
        self.__orientation = 0
        self.start = start
        self.end = end

        if start.y > end.y:
            self.__orientation = -1
        elif end.y > start.y:
            self.__orientation = 1

    @property
    def orientation(self):
        return self.__orientation

    def between_y_coords(self, y: float) -> bool:
        return (self.start.y < y and y < self.end.y) or (
            self.end.y < y and y < self.start.y
        )

    def horizontally_behind(self, x: float) -> bool:
        return x < min(self.start.x, self.end.x)


class Polygon2D:
    def __init__(self, *points: tuple[Iterable[float]]) -> None:
        if len(points) < 3:
            raise Exception("O polígono deve ter no mínimo 3 vértices.")
        
        self.vertices = [Point(*vertice) for vertice in points]
        self.edges = [
            Edge(vertice, self.vertices[(index + 1) % len(points)])
            for index, vertice in enumerate(self.vertices)
        ]

    def check_point_collide(self, point: Point) -> bool:
        laps = 0.0
        for edge in self.edges:
            if edge.horizontally_behind(point.x) and edge.between_y_coords(point.y):
                laps += edge.orientation
        return laps != 0

File: C#/Winding Number/Polygon.cs
namespace Polygon
{

    class Polygon2D
    {
        Point[] Vertices;
        Edge[] Edges;

        private IEnumerable<Point> MakePoints(float[][] points)
        {

            for (int i=0; i<points.Count(); i++)
            {
                yield return new Point(points[i]);
            }
        }

        private IEnumerable<Edge> MakeEdges(Point[] points)
        {
            for (int i=0; i<points.Count(); i++)
            {
                yield return new Edge(points[i], points[(i+1)%points.Count()]);
            }
        }

        public Polygon2D(float[][] points)
        {
            if (points.Count() < 3)
            {
                throw new Exception("O polígono deve ter no mínimo 3 vértices.");
            }
            Edges = MakeEdges(
                Vertices = MakePoints(points).ToArray()
            ).ToArray();
        }
    
        public bool CheckPointCollide(Point point)
        {
            int laps = 0;
            
            foreach(Edge edge in Edges)
            {
                if (edge.HorizontallyBehind(point.X) && edge.BetweenYCoords(point.Y))
                {
                    laps += edge.Orientation;
                }
            }
            return laps != 0;
        }

    }

    class Point
    {
        float[] Coord;

        public float X
        {
            get { return Coord[0]; }
            set { Coord[0]=value; }
        }
        public float Y
        {
            get { return Coord[1]; }
            set { Coord[1]=value; }
        }

        public Point(float[] coords)
        {
            Coord = coords;
        }

        public Point(float x, float y)
        {
            Coord = [x, y];
        }

    }

    class Edge{
        Point Start;
        Point End;
        private int _orientation = 0;
        public int Orientation
        {
            get {return _orientation;}
        }

        public Edge(Point start,Point end)
        {
            Start = start;
            End = end;

            if (start.Y > end.Y)
            {
                _orientation = -1;
            } else
            if (end.Y > start.Y)
            {
                _orientation = 1;
            }

        }

        public bool BetweenYCoords(float Y)
        {
            return (
                (Start.Y < Y && Y < End.Y) ||
                (End.Y < Y && Y < Start.Y)
            );
        }

        public bool HorizontallyBehind(float X)
        {
            return (
                X < Math.Min(Start.X, End.X)
            );
        }
    }

}
